### PR TITLE
tfsdk: Initial ImportResourceState support

### LINKED ISSUE #33: Support Managed Resource Import
### Module version

```
N/A
```

### Use-cases

Practitioners sometimes need to have Terraform adopt infrastructure that already exists, rather than creating the infrastructure with Terraform. Providers need to be able to support this, accepting a practitioner-provided ID and resource type and creating that resource in state. The framework should provide hooks for this.

### Attempted Solutions

N/A

### Proposal

This probably looks like an optional Import method that either Resources or ResourceTypes can implement (undecided on which, honestly).

### PR DESCRIPTION
**Pull request by bflad**
Closes #33

### CODE REVIEW

File: tfsdk/serve_import_test.go
Comment (bflad) [inline tfsdk/serve_import_test.go:77]: If this is not going to be resolved prior to v0.4.0, I suggest we at least update the ImportState method Go documentation to note that using resp.State.Set() is required for now.

File: tfsdk/request_import.go
Comment (paddycarver) [inline tfsdk/request_import.go:11]:
```suggestion
	// its own type of value and parsed during import. This value
```

File: tfsdk/request_import.go
Comment (paddycarver) [inline tfsdk/request_import.go:12]:
```suggestion
	// is not stored in the state unless the provider explicitly stores it.
```

File: tfsdk/serve_import_test.go
Comment (paddycarver) [reply tfsdk/serve_import_test.go:77]: I'd say we should prioritize it for 0.4.0.

File: tfsdk/serve_resource_config_validators_test.go
Comment (paddycarver) [inline tfsdk/serve_resource_config_validators_test.go:71]: Why do all of the other test resource types get implementations that update `importResourceStateCalledResourceType`, but the validate ones don't?

**Review by paddycarver — COMMENTED** [review]
Looks great to me, some small quibbles.

File: tfsdk/serve_resource_config_validators_test.go
Comment (bflad) [reply tfsdk/serve_resource_config_validators_test.go:71]: Ah, this is because when I implemented those two test resources I was trying to omit parts that would be uncalled so future travelers just focused on the importantly implemented methods. It should have been commented as such.

Do you have a preference between these options?

1. Add a comment in these implementations, e.g. `// Intentionally blank. Not expected to be called during testing.`
2. Copy the implementations of other test resources (recognizing they _should_ all stay in sync going forward; not sure how to enforce that)
3. Removing _other_ unexpected test resource implementations, doing something like option 1 so it is more clear

Thanks!

Comment (paddycarver) [reply tfsdk/serve_resource_config_validators_test.go:71]: I think my preference is 3, but replacing them with our not supported function and a comment saying that functionality is tested elsewhere. What do you think?

I mostly just am :/ about having code we don't _need_ in the test resources if we're not testing it, because that breeds confusion about whether it's used and where and what can be safely changed and what can't be in the future, I think. 🤔

Comment (bflad) [reply tfsdk/serve_resource_config_validators_test.go:71]: Agreed! That will work for the new `ImportState` method implementations and I'll add the comment to the other empty Create/Read/Update/Delete test implementations. Will update shortly after merging main to ensure this is ready to rock. 🎸

**Comment by bflad** [comment]
Everything should be ready for review again.

Something that gives me pause at the moment is how tooling will determine if a resource actually supports import and potentially how it can signal an example import identifier for documentation generation. I'm not sure if that should necessarily block an initial implementation, but it might be something worth iterating on in the future. Would love to hear thoughts and/or if I should create a followup issue.

**Comment by bflad** [comment]
Created #160 for potential followup.

**Review by paddycarver — APPROVED** [review]
Yeah, let's go ahead and do it. 👍 🚀 Nice work.

**Review by kmoe — APPROVED** [review]
LGTM

File: tfsdk/serve_import.go
Comment (paddycarver) [inline tfsdk/serve_import.go:85]: A thing worth thinking about here:

We need a resolution to #148 to make this work the way we want.

Another option is to have it do something like:

```go
vals := map[string]tftypes.Value{}
typ := resourceSchema.TerraformType(ctx)
for name, t := range typ.(tftypes.Object).AttributeTypes {
  vals[name] = tftypes.NewValue(t, nil)
}
emptyState := tftypes.NewValue(typ, vals)
```

This matches the more-expected (imho) default state, which is an object in state with nothing filled in, instead of nothing in state.

However, this is weird during failure cases. I'm not clear on whether Terraform will persist the object to state or write nothing to state in the face of an error diagnostic. If the former, we should default to an empty state and just rely on #148. If the latter, we should confirm that will continue to be the case, and maybe defaulting to the empty object (instead of the null object) is more expected? I dunno. I don't know that it makes a huge difference, just wanted to call it out here.

Comment (bflad) [reply tfsdk/serve_import.go:85]: The current logic in Terraform CLI is to return early on ImportResourceState error diagnostics, before saving state information:

https://github.com/hashicorp/terraform/blob/2afa0a5e75d76de7e47157114c579b3b1bff994f/internal/terraform/transform_import_state.go#L141-L154

Comment (paddycarver) [reply tfsdk/serve_import.go:85]: 🤔 how do we feel about just leaving it and waiting for feedback?

Comment (bflad) [reply tfsdk/serve_import.go:85]: We're going to merge this in as-is for now and iterate if necessary.

If we do wind up switching the implementation here in the future -- maybe creating the above code snippet as a method such as `(Schema).newState()` would be good to try and prevent "empty" object problems here and in any other potential places that might need it. That might be something we could export if/when multiple resource import support makes sense.

**Pull request merged**
